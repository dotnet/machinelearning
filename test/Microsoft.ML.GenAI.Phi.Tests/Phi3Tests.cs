// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Text;
using System.Text.Json;
using ApprovalTests;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using FluentAssertions;
using Microsoft.ML.GenAI.Core.Extension;
using Microsoft.ML.TestFramework;
using TorchSharp;
using Xunit;
using Xunit.Abstractions;
using static TorchSharp.torch;

namespace Microsoft.ML.GenAI.Phi.Tests;

public class Phi3Tests : BaseTestClass
{
    public Phi3Tests(ITestOutputHelper output) : base(output)
    {
        torch.set_default_device("meta");
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Mini4KShapeTest()
    {
        var modelWeightFolder = "C:\\Users\\xiaoyuz\\source\\repos\\Phi-3-mini-4k-instruct";
        var config = Path.Join(modelWeightFolder, "config.json");
        var modelConfig = JsonSerializer.Deserialize<Phi3Config>(File.ReadAllText(config)) ?? throw new ArgumentNullException(nameof(config));
        var model = new Phi3ForCasualLM(modelConfig);
        var stateDictStr = model.PeekShape();
        Approvals.Verify(stateDictStr);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Medium4KShapeTest()
    {
        var modelWeightFolder = "C:\\Users\\xiaoyuz\\source\\repos\\Phi-3-medium-4k-instruct";
        var config = Path.Join(modelWeightFolder, "config.json");
        var modelConfig = JsonSerializer.Deserialize<Phi3Config>(File.ReadAllText(config)) ?? throw new ArgumentNullException(nameof(config));
        var model = new Phi3ForCasualLM(modelConfig);
        var stateDictStr = model.PeekShape();
        Approvals.Verify(stateDictStr);
    }


    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Medium128KShapeTest()
    {
        var modelWeightFolder = "C:\\Users\\xiaoyuz\\source\\repos\\Phi-3-medium-128k-instruct";
        var config = Path.Join(modelWeightFolder, "config.json");
        var modelConfig = JsonSerializer.Deserialize<Phi3Config>(File.ReadAllText(config)) ?? throw new ArgumentNullException(nameof(config));
        var model = new Phi3ForCasualLM(modelConfig);
        var stateDictStr = model.PeekShape();
        Approvals.Verify(stateDictStr);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Mini128KShapeTest()
    {
        var modelWeightFolder = "C:\\Users\\xiaoyuz\\source\\repos\\Phi-3-mini-128k-instruct";
        var config = Path.Join(modelWeightFolder, "config.json");
        var modelConfig = JsonSerializer.Deserialize<Phi3Config>(File.ReadAllText(config)) ?? throw new ArgumentNullException(nameof(config));
        var model = new Phi3ForCasualLM(modelConfig);
        var stateDictStr = model.PeekShape();
        Approvals.Verify(stateDictStr);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Mini128KLayerSizeTest()
    {
        var dtype = ScalarType.BFloat16;
        var modelWeightFolder = "C:\\Users\\xiaoyuz\\source\\repos\\Phi-3-mini-128k-instruct";
        var config = Path.Join(modelWeightFolder, "config.json");
        var modelConfig = JsonSerializer.Deserialize<Phi3Config>(File.ReadAllText(config)) ?? throw new ArgumentNullException(nameof(config));
        modelConfig.DType = dtype;
        var model = new Phi3ForCasualLM(modelConfig);

        var size = model.GetSizeForEachDynamicLayerInBytes();
        // convert size to MB
        var sizeInMB = size.ToDictionary(x => x.Key, x => x.Value * 1.0f / 1024 / 1024);

        var json = JsonSerializer.Serialize(sizeInMB, new JsonSerializerOptions { WriteIndented = true });
        Approvals.Verify(json);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void TokenizerTest()
    {
        var modelWeightFolder = "C:\\Users\\xiaoyuz\\source\\repos\\Phi-3-mini-4k-instruct";
        var tokenizer = LLama2Tokenizer.FromPretrained(modelWeightFolder);
        tokenizer.BosId.Should().Be(1);
        tokenizer.EosId.Should().Be(2);
        var messages = new string[]
        {
            "Can you provide ways to eat combinations of bananas and dragonfruits?",
            "Sure! Here are some ways to eat bananas and dragonfruits together: 1. Banana and dragonfruit smoothie: Blend bananas and dragonfruits together with some milk and honey. 2. Banana and dragonfruit salad: Mix sliced bananas and dragonfruits together with some lemon juice and honey.",
            "What about solving an 2x + 3 = 7 equation?",
            "\nCount to 3\n",
            "<|user|>\nCount to 3<|end|>\n<|assistant|>",
        };
        var sb = new StringBuilder();
        foreach (var message in messages)
        {
            var tokenized = tokenizer.Encode(message, true, false);
            var tokenizedStr = string.Join(", ", tokenized.Select(x => x.ToString()));

            sb.AppendLine(tokenizedStr);
        }
        Approvals.Verify(sb.ToString());
    }
}
