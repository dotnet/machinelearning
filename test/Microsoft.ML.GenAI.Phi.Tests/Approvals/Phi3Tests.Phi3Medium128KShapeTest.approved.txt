0: lm_head.weight shape: [32064, 5120]
1: model.embed_tokens.weight shape: [32064, 5120]
2: model.layers.0.input_layernorm.weight shape: [5120]
3: model.layers.0.mlp.down_proj.weight shape: [5120, 17920]
4: model.layers.0.mlp.gate_up_proj.weight shape: [35840, 5120]
5: model.layers.0.post_attention_layernorm.weight shape: [5120]
6: model.layers.0.self_attn.o_proj.weight shape: [5120, 5120]
7: model.layers.0.self_attn.qkv_proj.weight shape: [7680, 5120]
8: model.layers.1.input_layernorm.weight shape: [5120]
9: model.layers.1.mlp.down_proj.weight shape: [5120, 17920]
10: model.layers.1.mlp.gate_up_proj.weight shape: [35840, 5120]
11: model.layers.1.post_attention_layernorm.weight shape: [5120]
12: model.layers.1.self_attn.o_proj.weight shape: [5120, 5120]
13: model.layers.1.self_attn.qkv_proj.weight shape: [7680, 5120]
14: model.layers.10.input_layernorm.weight shape: [5120]
15: model.layers.10.mlp.down_proj.weight shape: [5120, 17920]
16: model.layers.10.mlp.gate_up_proj.weight shape: [35840, 5120]
17: model.layers.10.post_attention_layernorm.weight shape: [5120]
18: model.layers.10.self_attn.o_proj.weight shape: [5120, 5120]
19: model.layers.10.self_attn.qkv_proj.weight shape: [7680, 5120]
20: model.layers.11.input_layernorm.weight shape: [5120]
21: model.layers.11.mlp.down_proj.weight shape: [5120, 17920]
22: model.layers.11.mlp.gate_up_proj.weight shape: [35840, 5120]
23: model.layers.11.post_attention_layernorm.weight shape: [5120]
24: model.layers.11.self_attn.o_proj.weight shape: [5120, 5120]
25: model.layers.11.self_attn.qkv_proj.weight shape: [7680, 5120]
26: model.layers.12.input_layernorm.weight shape: [5120]
27: model.layers.12.mlp.down_proj.weight shape: [5120, 17920]
28: model.layers.12.mlp.gate_up_proj.weight shape: [35840, 5120]
29: model.layers.12.post_attention_layernorm.weight shape: [5120]
30: model.layers.12.self_attn.o_proj.weight shape: [5120, 5120]
31: model.layers.12.self_attn.qkv_proj.weight shape: [7680, 5120]
32: model.layers.13.input_layernorm.weight shape: [5120]
33: model.layers.13.mlp.down_proj.weight shape: [5120, 17920]
34: model.layers.13.mlp.gate_up_proj.weight shape: [35840, 5120]
35: model.layers.13.post_attention_layernorm.weight shape: [5120]
36: model.layers.13.self_attn.o_proj.weight shape: [5120, 5120]
37: model.layers.13.self_attn.qkv_proj.weight shape: [7680, 5120]
38: model.layers.14.input_layernorm.weight shape: [5120]
39: model.layers.14.mlp.down_proj.weight shape: [5120, 17920]
40: model.layers.14.mlp.gate_up_proj.weight shape: [35840, 5120]
41: model.layers.14.post_attention_layernorm.weight shape: [5120]
42: model.layers.14.self_attn.o_proj.weight shape: [5120, 5120]
43: model.layers.14.self_attn.qkv_proj.weight shape: [7680, 5120]
44: model.layers.15.input_layernorm.weight shape: [5120]
45: model.layers.15.mlp.down_proj.weight shape: [5120, 17920]
46: model.layers.15.mlp.gate_up_proj.weight shape: [35840, 5120]
47: model.layers.15.post_attention_layernorm.weight shape: [5120]
48: model.layers.15.self_attn.o_proj.weight shape: [5120, 5120]
49: model.layers.15.self_attn.qkv_proj.weight shape: [7680, 5120]
50: model.layers.16.input_layernorm.weight shape: [5120]
51: model.layers.16.mlp.down_proj.weight shape: [5120, 17920]
52: model.layers.16.mlp.gate_up_proj.weight shape: [35840, 5120]
53: model.layers.16.post_attention_layernorm.weight shape: [5120]
54: model.layers.16.self_attn.o_proj.weight shape: [5120, 5120]
55: model.layers.16.self_attn.qkv_proj.weight shape: [7680, 5120]
56: model.layers.17.input_layernorm.weight shape: [5120]
57: model.layers.17.mlp.down_proj.weight shape: [5120, 17920]
58: model.layers.17.mlp.gate_up_proj.weight shape: [35840, 5120]
59: model.layers.17.post_attention_layernorm.weight shape: [5120]
60: model.layers.17.self_attn.o_proj.weight shape: [5120, 5120]
61: model.layers.17.self_attn.qkv_proj.weight shape: [7680, 5120]
62: model.layers.18.input_layernorm.weight shape: [5120]
63: model.layers.18.mlp.down_proj.weight shape: [5120, 17920]
64: model.layers.18.mlp.gate_up_proj.weight shape: [35840, 5120]
65: model.layers.18.post_attention_layernorm.weight shape: [5120]
66: model.layers.18.self_attn.o_proj.weight shape: [5120, 5120]
67: model.layers.18.self_attn.qkv_proj.weight shape: [7680, 5120]
68: model.layers.19.input_layernorm.weight shape: [5120]
69: model.layers.19.mlp.down_proj.weight shape: [5120, 17920]
70: model.layers.19.mlp.gate_up_proj.weight shape: [35840, 5120]
71: model.layers.19.post_attention_layernorm.weight shape: [5120]
72: model.layers.19.self_attn.o_proj.weight shape: [5120, 5120]
73: model.layers.19.self_attn.qkv_proj.weight shape: [7680, 5120]
74: model.layers.2.input_layernorm.weight shape: [5120]
75: model.layers.2.mlp.down_proj.weight shape: [5120, 17920]
76: model.layers.2.mlp.gate_up_proj.weight shape: [35840, 5120]
77: model.layers.2.post_attention_layernorm.weight shape: [5120]
78: model.layers.2.self_attn.o_proj.weight shape: [5120, 5120]
79: model.layers.2.self_attn.qkv_proj.weight shape: [7680, 5120]
80: model.layers.20.input_layernorm.weight shape: [5120]
81: model.layers.20.mlp.down_proj.weight shape: [5120, 17920]
82: model.layers.20.mlp.gate_up_proj.weight shape: [35840, 5120]
83: model.layers.20.post_attention_layernorm.weight shape: [5120]
84: model.layers.20.self_attn.o_proj.weight shape: [5120, 5120]
85: model.layers.20.self_attn.qkv_proj.weight shape: [7680, 5120]
86: model.layers.21.input_layernorm.weight shape: [5120]
87: model.layers.21.mlp.down_proj.weight shape: [5120, 17920]
88: model.layers.21.mlp.gate_up_proj.weight shape: [35840, 5120]
89: model.layers.21.post_attention_layernorm.weight shape: [5120]
90: model.layers.21.self_attn.o_proj.weight shape: [5120, 5120]
91: model.layers.21.self_attn.qkv_proj.weight shape: [7680, 5120]
92: model.layers.22.input_layernorm.weight shape: [5120]
93: model.layers.22.mlp.down_proj.weight shape: [5120, 17920]
94: model.layers.22.mlp.gate_up_proj.weight shape: [35840, 5120]
95: model.layers.22.post_attention_layernorm.weight shape: [5120]
96: model.layers.22.self_attn.o_proj.weight shape: [5120, 5120]
97: model.layers.22.self_attn.qkv_proj.weight shape: [7680, 5120]
98: model.layers.23.input_layernorm.weight shape: [5120]
99: model.layers.23.mlp.down_proj.weight shape: [5120, 17920]
100: model.layers.23.mlp.gate_up_proj.weight shape: [35840, 5120]
101: model.layers.23.post_attention_layernorm.weight shape: [5120]
102: model.layers.23.self_attn.o_proj.weight shape: [5120, 5120]
103: model.layers.23.self_attn.qkv_proj.weight shape: [7680, 5120]
104: model.layers.24.input_layernorm.weight shape: [5120]
105: model.layers.24.mlp.down_proj.weight shape: [5120, 17920]
106: model.layers.24.mlp.gate_up_proj.weight shape: [35840, 5120]
107: model.layers.24.post_attention_layernorm.weight shape: [5120]
108: model.layers.24.self_attn.o_proj.weight shape: [5120, 5120]
109: model.layers.24.self_attn.qkv_proj.weight shape: [7680, 5120]
110: model.layers.25.input_layernorm.weight shape: [5120]
111: model.layers.25.mlp.down_proj.weight shape: [5120, 17920]
112: model.layers.25.mlp.gate_up_proj.weight shape: [35840, 5120]
113: model.layers.25.post_attention_layernorm.weight shape: [5120]
114: model.layers.25.self_attn.o_proj.weight shape: [5120, 5120]
115: model.layers.25.self_attn.qkv_proj.weight shape: [7680, 5120]
116: model.layers.26.input_layernorm.weight shape: [5120]
117: model.layers.26.mlp.down_proj.weight shape: [5120, 17920]
118: model.layers.26.mlp.gate_up_proj.weight shape: [35840, 5120]
119: model.layers.26.post_attention_layernorm.weight shape: [5120]
120: model.layers.26.self_attn.o_proj.weight shape: [5120, 5120]
121: model.layers.26.self_attn.qkv_proj.weight shape: [7680, 5120]
122: model.layers.27.input_layernorm.weight shape: [5120]
123: model.layers.27.mlp.down_proj.weight shape: [5120, 17920]
124: model.layers.27.mlp.gate_up_proj.weight shape: [35840, 5120]
125: model.layers.27.post_attention_layernorm.weight shape: [5120]
126: model.layers.27.self_attn.o_proj.weight shape: [5120, 5120]
127: model.layers.27.self_attn.qkv_proj.weight shape: [7680, 5120]
128: model.layers.28.input_layernorm.weight shape: [5120]
129: model.layers.28.mlp.down_proj.weight shape: [5120, 17920]
130: model.layers.28.mlp.gate_up_proj.weight shape: [35840, 5120]
131: model.layers.28.post_attention_layernorm.weight shape: [5120]
132: model.layers.28.self_attn.o_proj.weight shape: [5120, 5120]
133: model.layers.28.self_attn.qkv_proj.weight shape: [7680, 5120]
134: model.layers.29.input_layernorm.weight shape: [5120]
135: model.layers.29.mlp.down_proj.weight shape: [5120, 17920]
136: model.layers.29.mlp.gate_up_proj.weight shape: [35840, 5120]
137: model.layers.29.post_attention_layernorm.weight shape: [5120]
138: model.layers.29.self_attn.o_proj.weight shape: [5120, 5120]
139: model.layers.29.self_attn.qkv_proj.weight shape: [7680, 5120]
140: model.layers.3.input_layernorm.weight shape: [5120]
141: model.layers.3.mlp.down_proj.weight shape: [5120, 17920]
142: model.layers.3.mlp.gate_up_proj.weight shape: [35840, 5120]
143: model.layers.3.post_attention_layernorm.weight shape: [5120]
144: model.layers.3.self_attn.o_proj.weight shape: [5120, 5120]
145: model.layers.3.self_attn.qkv_proj.weight shape: [7680, 5120]
146: model.layers.30.input_layernorm.weight shape: [5120]
147: model.layers.30.mlp.down_proj.weight shape: [5120, 17920]
148: model.layers.30.mlp.gate_up_proj.weight shape: [35840, 5120]
149: model.layers.30.post_attention_layernorm.weight shape: [5120]
150: model.layers.30.self_attn.o_proj.weight shape: [5120, 5120]
151: model.layers.30.self_attn.qkv_proj.weight shape: [7680, 5120]
152: model.layers.31.input_layernorm.weight shape: [5120]
153: model.layers.31.mlp.down_proj.weight shape: [5120, 17920]
154: model.layers.31.mlp.gate_up_proj.weight shape: [35840, 5120]
155: model.layers.31.post_attention_layernorm.weight shape: [5120]
156: model.layers.31.self_attn.o_proj.weight shape: [5120, 5120]
157: model.layers.31.self_attn.qkv_proj.weight shape: [7680, 5120]
158: model.layers.32.input_layernorm.weight shape: [5120]
159: model.layers.32.mlp.down_proj.weight shape: [5120, 17920]
160: model.layers.32.mlp.gate_up_proj.weight shape: [35840, 5120]
161: model.layers.32.post_attention_layernorm.weight shape: [5120]
162: model.layers.32.self_attn.o_proj.weight shape: [5120, 5120]
163: model.layers.32.self_attn.qkv_proj.weight shape: [7680, 5120]
164: model.layers.33.input_layernorm.weight shape: [5120]
165: model.layers.33.mlp.down_proj.weight shape: [5120, 17920]
166: model.layers.33.mlp.gate_up_proj.weight shape: [35840, 5120]
167: model.layers.33.post_attention_layernorm.weight shape: [5120]
168: model.layers.33.self_attn.o_proj.weight shape: [5120, 5120]
169: model.layers.33.self_attn.qkv_proj.weight shape: [7680, 5120]
170: model.layers.34.input_layernorm.weight shape: [5120]
171: model.layers.34.mlp.down_proj.weight shape: [5120, 17920]
172: model.layers.34.mlp.gate_up_proj.weight shape: [35840, 5120]
173: model.layers.34.post_attention_layernorm.weight shape: [5120]
174: model.layers.34.self_attn.o_proj.weight shape: [5120, 5120]
175: model.layers.34.self_attn.qkv_proj.weight shape: [7680, 5120]
176: model.layers.35.input_layernorm.weight shape: [5120]
177: model.layers.35.mlp.down_proj.weight shape: [5120, 17920]
178: model.layers.35.mlp.gate_up_proj.weight shape: [35840, 5120]
179: model.layers.35.post_attention_layernorm.weight shape: [5120]
180: model.layers.35.self_attn.o_proj.weight shape: [5120, 5120]
181: model.layers.35.self_attn.qkv_proj.weight shape: [7680, 5120]
182: model.layers.36.input_layernorm.weight shape: [5120]
183: model.layers.36.mlp.down_proj.weight shape: [5120, 17920]
184: model.layers.36.mlp.gate_up_proj.weight shape: [35840, 5120]
185: model.layers.36.post_attention_layernorm.weight shape: [5120]
186: model.layers.36.self_attn.o_proj.weight shape: [5120, 5120]
187: model.layers.36.self_attn.qkv_proj.weight shape: [7680, 5120]
188: model.layers.37.input_layernorm.weight shape: [5120]
189: model.layers.37.mlp.down_proj.weight shape: [5120, 17920]
190: model.layers.37.mlp.gate_up_proj.weight shape: [35840, 5120]
191: model.layers.37.post_attention_layernorm.weight shape: [5120]
192: model.layers.37.self_attn.o_proj.weight shape: [5120, 5120]
193: model.layers.37.self_attn.qkv_proj.weight shape: [7680, 5120]
194: model.layers.38.input_layernorm.weight shape: [5120]
195: model.layers.38.mlp.down_proj.weight shape: [5120, 17920]
196: model.layers.38.mlp.gate_up_proj.weight shape: [35840, 5120]
197: model.layers.38.post_attention_layernorm.weight shape: [5120]
198: model.layers.38.self_attn.o_proj.weight shape: [5120, 5120]
199: model.layers.38.self_attn.qkv_proj.weight shape: [7680, 5120]
200: model.layers.39.input_layernorm.weight shape: [5120]
201: model.layers.39.mlp.down_proj.weight shape: [5120, 17920]
202: model.layers.39.mlp.gate_up_proj.weight shape: [35840, 5120]
203: model.layers.39.post_attention_layernorm.weight shape: [5120]
204: model.layers.39.self_attn.o_proj.weight shape: [5120, 5120]
205: model.layers.39.self_attn.qkv_proj.weight shape: [7680, 5120]
206: model.layers.4.input_layernorm.weight shape: [5120]
207: model.layers.4.mlp.down_proj.weight shape: [5120, 17920]
208: model.layers.4.mlp.gate_up_proj.weight shape: [35840, 5120]
209: model.layers.4.post_attention_layernorm.weight shape: [5120]
210: model.layers.4.self_attn.o_proj.weight shape: [5120, 5120]
211: model.layers.4.self_attn.qkv_proj.weight shape: [7680, 5120]
212: model.layers.5.input_layernorm.weight shape: [5120]
213: model.layers.5.mlp.down_proj.weight shape: [5120, 17920]
214: model.layers.5.mlp.gate_up_proj.weight shape: [35840, 5120]
215: model.layers.5.post_attention_layernorm.weight shape: [5120]
216: model.layers.5.self_attn.o_proj.weight shape: [5120, 5120]
217: model.layers.5.self_attn.qkv_proj.weight shape: [7680, 5120]
218: model.layers.6.input_layernorm.weight shape: [5120]
219: model.layers.6.mlp.down_proj.weight shape: [5120, 17920]
220: model.layers.6.mlp.gate_up_proj.weight shape: [35840, 5120]
221: model.layers.6.post_attention_layernorm.weight shape: [5120]
222: model.layers.6.self_attn.o_proj.weight shape: [5120, 5120]
223: model.layers.6.self_attn.qkv_proj.weight shape: [7680, 5120]
224: model.layers.7.input_layernorm.weight shape: [5120]
225: model.layers.7.mlp.down_proj.weight shape: [5120, 17920]
226: model.layers.7.mlp.gate_up_proj.weight shape: [35840, 5120]
227: model.layers.7.post_attention_layernorm.weight shape: [5120]
228: model.layers.7.self_attn.o_proj.weight shape: [5120, 5120]
229: model.layers.7.self_attn.qkv_proj.weight shape: [7680, 5120]
230: model.layers.8.input_layernorm.weight shape: [5120]
231: model.layers.8.mlp.down_proj.weight shape: [5120, 17920]
232: model.layers.8.mlp.gate_up_proj.weight shape: [35840, 5120]
233: model.layers.8.post_attention_layernorm.weight shape: [5120]
234: model.layers.8.self_attn.o_proj.weight shape: [5120, 5120]
235: model.layers.8.self_attn.qkv_proj.weight shape: [7680, 5120]
236: model.layers.9.input_layernorm.weight shape: [5120]
237: model.layers.9.mlp.down_proj.weight shape: [5120, 17920]
238: model.layers.9.mlp.gate_up_proj.weight shape: [35840, 5120]
239: model.layers.9.post_attention_layernorm.weight shape: [5120]
240: model.layers.9.self_attn.o_proj.weight shape: [5120, 5120]
241: model.layers.9.self_attn.qkv_proj.weight shape: [7680, 5120]
242: model.norm.weight shape: [5120]
